web3.js/src/connection.ts

```diff
@@ -4281,75 +4232,237 @@ export class Connection {
-        this._buildArgs([sub.publicKey], sub.commitment, 'base64'),
-      );
-    }
-
-    for (let id of programKeys) {
-      const sub = this._programAccountChangeSubscriptions[id];
-      this._subscribe(
-        sub,
-        'programSubscribe',
-        this._buildArgs([sub.programId], sub.commitment, 'base64', {
-          filters: sub.filters,
-        }),
-      );
-    }
-
-    for (let id of slotKeys) {
-      const sub = this._slotSubscriptions[id];
-      this._subscribe(sub, 'slotSubscribe', []);
-    }
-
-    for (let id of slotUpdateKeys) {
-      const sub = this._slotUpdateSubscriptions[id];
-      this._subscribe(sub, 'slotsUpdatesSubscribe', []);
-    }
-
-    for (let id of signatureKeys) {
-      const sub = this._signatureSubscriptions[id];
-      const args: any[] = [sub.signature];
-      if (sub.options) args.push(sub.options);
-      this._subscribe(sub, 'signatureSubscribe', args);
-    }
+    await Promise.all(
+      // Don't be tempted to change this to `Object.entries`. We call
+      // `_updateSubscriptions` recursively when processing the state,
+      // so it's important that we look up the *current* version of
+      // each subscription, every time we process a hash.
+      Object.keys(this._subscriptionsByHash).map(async hash => {
+        const subscription = this._subscriptionsByHash[hash];
+        if (subscription === undefined) {
+          // This entry has since been deleted. Skip.
+          return;
+        }
+        switch (subscription.state) {
+          case 'pending':
+          case 'unsubscribed':
+            if (subscription.callbacks.size === 0) {
+              /**
+               * You can end up here when:
+               *
+               * - a subscription has recently unsubscribed
+               *   without having new callbacks added to it
+               *   while the unsubscribe was in flight, or
+               * - when a pending subscription has its
+               *   listeners removed before a request was
+               *   sent to the server.
+               *
+               * Being that nobody is interested in this
+               * subscription any longer, delete it.
+               */
+              delete this._subscriptionsByHash[hash];
+              if (subscription.state === 'unsubscribed') {
+                delete this._subscriptionCallbacksByServerSubscriptionId[
+                  subscription.serverSubscriptionId
+                ];
+              }
+              await this._updateSubscriptions();
+              return;
+            }
+            await (async () => {
+              const {args, method} = subscription;
+              try {
+                this._subscriptionsByHash[hash] = {
+                  ...subscription,
+                  state: 'subscribing',
+                };
+                const serverSubscriptionId: ServerSubscriptionId =
+                  (await this._rpcWebSocket.call(method, args)) as number;
```

Comment:
While calling this method, the websocket could be closed and all subscriptions are set to `'pending'`, I think we should check for that before setting to `'subscribed'` here (this applies to unsubscribe as well).

Reply:
Woo, that sent me down a debugging rabbit hole. OK, I got to the bottom of this. Wow. From MDN.

> The browser will throw an exception if you call send() when the connection is in the CONNECTING state. If you call send() when the connection is in the CLOSING or CLOSED states, the browser will silently discard the data.

I just presumed that if you invoke `call()` when the socket is down, it would _throw an exception_ and process the catch block below that sets the subscription back to `pending`.

The ‘bare metal’ socket doesn't throw. It just drops the data, and never calls the callback.

```
ws = new WebSocket('wss://demo.piesocket.com/v3/channel_1?api_key=VCXCEuvhGcBDP7XhiJJUDvR1e1D3eiVjgZ9VRiaV&notify_self')
ws.onopen = () => {
    ws.close();
    console.log('is `CLOSING`', ws.readyState === ws.CLOSING);
    try {
        ws.send(JSON.stringify('hello'));
        console.log('sent, when I expected an error to be thrown');
    } catch(e) {
        console.error('legit caught an error', e);
    }
};
```

**But**, the rpc-websockets client _always_ throws if it doesn't receive a reply within a timeout.

https://github.com/elpheria/rpc-websockets/blob/trunk/src/lib/client.ts#L157-L172

That's the bug here. The `try` block will only ever continue past `await this._rpcWebSocket.call()` if a reply was received, in which case we're all good. The `catch` block however _might_ get called because of a disconnection-caused timeout. It's _there_ where I need to make sure that we're still in the `subscribing` state before I just hard shunt the subscription back to `pending`. And _that_ I can write a test for.

Reply:
Alright @jstarry, one more commit on top: ‘fix: bail on async…’ should fix this.

Reply:
A socket could disconnect while we're literally in the middle of processing the reply, 100%. The thing is that the disconnect _happens_ in a thread outside the JS engine (ie. in the browser's networking stack) so there's no possible way for it to show up on the single JavaScript thread until the next runloop.

The async `_rpcWebSocket.call()` can only resume in one of two ways:

* throw (reject) when the reply timeout gets hit.
* resume execution (resolve) because it received a reply.

Presume, now, that `_rpcWebSocket.call()` has yielded to the scheduler and is waiting for one of those two things to happen.

* If a close event happens before we get a reply, the only possible thing that could happen next is for the reply timeout to be hit†, which would trigger our `catch` which handles this properly now.
* If a reply comes in, then we're definitely not closed, and every statement after the `await` will run synchronously until there are no more statements to evaluate.

Now, imagine that a close happened, in a _real world_ sense, while that code after the `await` was running (eg. I yanked out the network cable). The _browser_ will learn of the close, for sure, but there's literally no way for the browser to schedule the close event on the JS thread until _after_ the reply is done being processed. The browser can fire the `close` event into the JS thread all it wants, but it's not going to be able to preempt the code after the `await`.

```
this._subscriptionsByHash[hash] = {
  ...subscription,
  serverSubscriptionId,
  state: 'subscribed',
};
this._subscriptionCallbacksByServerSubscriptionId[
  serverSubscriptionId
] = subscription.callbacks;
await this._updateSubscriptions();
```

…by sneaking a call to `_wsOnClose` in between those three lines.

JavaScript being single threaded is one of the worst things about JavaScript, but also definitely one of the best things about JavaScript.

---
† If it's literally possible to _receive_ and _process_ a data frame (reply) _after_ having received a `close` event on the JS thread, then everything I said above doesn't matter, I will eat my hat, and websockets are terrible.

Reply:
My understanding from [this gist](https://gist.github.com/jstarry/9b86287eec9468696d5dd28aa7a6c50a) is that even if an async rpc call receives and processes a reply, we might not pick up where we left off until after processing other events in the work queue

Reply:
Oh totally, but other events will mostly touch _other_ subscriptions, and the ones that do touch this subscription will perform no actions because it's in the `'subscribing'` state:

```
case 'subscribing':
case 'unsubscribing':
  break;
```

Receiving a reply corresponds to [this event firing](https://github.com/elpheria/rpc-websockets/blob/trunk/src/lib/client.ts#L306) in the `rpc-websockets` library:

```
this.socket.addEventListener("message", ({data: message}) => { ... }
```

All of the code of that event handler fires sync, including [this bit at the end](https://github.com/elpheria/rpc-websockets/blob/trunk/src/lib/client.ts#L360-L363):

```
if (message.error)
  this.queue[message.id].promise[1](message.error)
else
  this.queue[message.id].promise[0](message.result)
```

`promise[0]` there corresponds to the `resolve` of `this._rpcWebSockets.call(...)`. When it's called with `message.result` (the reply) then all of the code after _our_ `await` will run synchronously:

```
this._subscriptionsByHash[hash] = {
  ...subscription,
  serverSubscriptionId,
  state: 'subscribed',
};
this._subscriptionCallbacksByServerSubscriptionId[
  serverSubscriptionId
] = subscription.callbacks;
await this._updateSubscriptions();
```

So essentially you can draw a straight line between the `'message'` event firing and all of that code running before yielding back to the browser. This means that the act of receiving a reply implies two things:

* you were definitely not `CLOSED` or `CLOSING` at that moment (otherwise you wouldn't have been able to receive `'message'`), and
* no `'close'` event can sneak into the JS thread while you're running the reply processing code.

Take a look at these branches of `connection-subscriptions.test.ts`:

```
Subscriptions
  The `accountSubscribe` RPC method
    attaching the first notification listener
      ✓ results in a subscription request being made to the RPC
      then unsubscribing that listener before the subscription has been acknowledged by the server
        once the subscription has been acknowledged by the server
          ✓ results in the subscription being torn down immediately
      then having the socket connection drop unexpectedly
        then unsubscribing that listener
          upon the socket connection reopening
            ✓ does not result in a new subscription request being made to the RPC
        upon the socket connection reopening
          ✓ results in a new subscription request being made to the RPC
          then upon the prior subscription fataling (eg. because its timeout triggers)
            ✓ does not result in a new subscription request being made to the RPC
```

Those exercise what happens when you have an in-flight `'subscribing'` state and one of two things happens:

* the user decides they don't care anymore and call `remove*`
* the connection drops

Thoughts? :)

Reply:
> So essentially you can draw a straight line between the 'message' event firing and all of that code running before yielding back to the browser. This means that the act of receiving a reply implies two things:

Ah, this makes a lot of sense. Looks good then, thanks :)